website_product_visibility_control_timer/models/product_template.py

```diff
@@ -0,0 +1,31 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    expected_publish_date = fields.Datetime()
+    expected_unpublish_date = fields.Datetime()
+
+    def _check_website_product_visibility(self):
+        publish_products = self.env["product.template"].search(
+            [
+                ("website_published", "=", False),
+                ("expected_publish_date", "<=", fields.datetime.now()),
+                "|",
+                ("expected_unpublish_date", ">", fields.datetime.now()),
+                ("expected_unpublish_date", "=", False),
+            ]
```

Comment:
This domain does not work if user intends to first unpublish an already published product, and then unpublish it.  I know this scenario sounds rather unlikely, but can we handle this case also?